### PR TITLE
JAMES-2134 MDN Report parser

### PR DESCRIPTION
**Pull request by chibenwa**
I know that hand written parser can look scary, hence I put a lot of attention of writting details edge cases and many tests (cloc reports 633 loc for main code and 1240 for tests).

Also the architecture of the parser is simple:

 - Step 1: Unwrap lines
 - Step 2: Compute fields
 - Step 3: parse each field
 - Step 4: Allow to retrieve a specific field value
 - Step 5: Generate the specific MDN

I took the desision to limit String parsing to the use of **indexOf** (getting delimiter). Anyway, even if detailed implementation don't convince you, we still get:

 - Parser struture with a working implementation
 - Detailed unit tests for each components

Of course this is implemented with a LENIENT and a STRICT mode.

Hope that it makes you not too scary...

### CODE REVIEW

File: mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    public static class UnwrapAccumulator {
+        private final ImmutableList<String> unwrappedLines;
+        private final Optional<String> currentUnwrapedLine;
+
+        public UnwrapAccumulator() {
+            this(ImmutableList.of(), Optional.empty());
+        }
+
+        private UnwrapAccumulator(ImmutableList<String> unwrappedLine, Optional<String> currentUnwrapedLine) {
+            this.unwrappedLines = unwrappedLine;
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java]: unwrappedLine**s**

File: mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                    .add(currentUnwrapedLine.get())
+                    .build();
+            }
+            return unwrappedLines;
+        }
+
+        public UnwrapAccumulator addLine(String line) {
+            boolean isWrapped = line.startsWith(" ");
+            if (isWrapped) {
+                String newCurrentWrappedLine = currentUnwrapedLine.map(s -> s + "\n").orElse("") + line.substring(1);
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java]: 1 is a bit magic. It's `" ".length` if i'm not wrong

File: mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    public static class UnwrapAccumulator {
+        private final ImmutableList<String> unwrappedLines;
+        private final Optional<String> currentUnwrapedLine;
+
+        public UnwrapAccumulator() {
+            this(ImmutableList.of(), Optional.empty());
+        }
+
+        private UnwrapAccumulator(ImmutableList<String> unwrappedLine, Optional<String> currentUnwrapedLine) {
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java]: wrapped and wraped, you have to choose

File: mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+public class Unwrapper {
+    public static ImmutableList<String> unwrap(String initialText) {
+        Preconditions.checkNotNull(initialText);
+
+        List<String> lines = Splitter.on("\r\n")
+            .splitToList(initialText);
+
+        return lines.stream()
+            .reduce(new UnwrapAccumulator(), UnwrapAccumulator::addLine, UnwrapAccumulator::combine)
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java]: how are you sure combine will never be called ?

File: mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            return unwrappedLines;
+        }
+
+        public UnwrapAccumulator addLine(String line) {
+            boolean isWrapped = line.startsWith(" ");
+            if (isWrapped) {
+                String newCurrentWrappedLine = currentUnwrapedLine.map(s -> s + "\n").orElse("") + line.substring(1);
+                return new UnwrapAccumulator(unwrappedLines, Optional.of(newCurrentWrappedLine));
+            }
+            return new UnwrapAccumulator(getUnwrappedLines(), Optional.of(line));
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java]: it's not clear why in this case you call getUnwrappedLines and two lines before you use the unwrappedLines field directly ?

File: mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            if (currentUnwrapedLine.isPresent()) {
+                return ImmutableList.<String>builder()
+                    .addAll(unwrappedLines)
+                    .add(currentUnwrapedLine.get())
+                    .build();
+            }
+            return unwrappedLines;
+        }
+
+        public UnwrapAccumulator addLine(String line) {
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java]: I think public doesn't make sense on a private static class.

File: mdn/src/test/java/org/apache/james/mdn/parsing/UnwrapperTest.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+package org.apache.james.mdn.parsing;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class UnwrapperTest {
```
Comment (mbaechler) [inline mdn/src/test/java/org/apache/james/mdn/parsing/UnwrapperTest.java]: what if you start with a single space line ?
what if you have several single space lines ?

File: mdn/src/main/java/org/apache/james/mdn/parsing/UnparsedField.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+import java.util.Objects;
+import java.util.Optional;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+public class UnparsedField {
+
+    public static class Factory {
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/UnparsedField.java]: A class that have a single parse method is probably more a Parser then a Factory

File: mdn/src/main/java/org/apache/james/mdn/parsing/UnparsedField.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+            String name = rawField.substring(0, nameDelimiterPosition);
+            String value = rawField.substring(nameDelimiterPosition + 1);
+            if (name.contains("\n")) {
+                return handleError("Field name can not contain line breaks.");
+            }
+            return Optional.of(new UnparsedField(name, value));
+        }
+
+        private Optional<UnparsedField> handleError(String errorMessage) {
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/UnparsedField.java]: That's exactly why people created Either or Validation : let the client decide what to do with errors.

File: mdn/src/main/java/org/apache/james/mdn/parsing/UnparsedField.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+public class UnparsedField {
+
+    public static class Factory {
+        private final boolean strict;
+
+        public Factory(boolean strict) {
+            this.strict = strict;
+        }
+
+        public Optional<UnparsedField> parse(String rawField) {
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/UnparsedField.java]: It's funny that your parser return an Unparse-something.

File: mdn/src/test/java/org/apache/james/mdn/parsing/UnparsedFieldTest.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import nl.jqno.equalsverifier.EqualsVerifier;
+
+public class UnparsedFieldTest {
+    @Rule
+    public ExpectedException expectedException = ExpectedException.none();
```
Comment (mbaechler) [inline mdn/src/test/java/org/apache/james/mdn/parsing/UnparsedFieldTest.java]: I think there's no point using that anymore now that everything is java8, assertj provides `assertThatThrownBy`

File: mdn/src/main/java/org/apache/james/mdn/parsing/DispositionParser.java [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    public Optional<Field> parse(String value) {
+        Preconditions.checkNotNull(value);
+
+        int slashPosition = value.indexOf('/');
+        Optional<DispositionActionMode> actionMode = parseActionMode(value, slashPosition);
+        String afterActionMode = value.substring(slashPosition + 1);
+
+        int semiColonPosition = afterActionMode.indexOf(';');
+        Optional<DispositionSendingMode> sendingMode = parseSendingMode(afterActionMode, semiColonPosition);
+        String afterSendingMode = afterActionMode.substring(semiColonPosition + 1);
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/DispositionParser.java]: you're not consistent about where you split the string : once in submethod, once in the body

File: mdn/src/main/java/org/apache/james/mdn/parsing/DispositionParser.java [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+import org.apache.james.mdn.modifier.DispositionModifier;
+import org.apache.james.mdn.sending.mode.DispositionSendingMode;
+import org.apache.james.mdn.type.DispositionType;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+
+public class DispositionParser implements FieldsParser.FieldParser {
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/parsing/DispositionParser.java]: having to manage strict/not-strict make everything so much longer, it's definitely not the way to go IMO.

File: mdn/src/test/java/org/apache/james/mdn/parsing/ReportingUserAgentParserTest.java [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+    public void parseStrictShouldAllowNoProduct() {
+        assertThat(strict.parse("ip"))
+            .contains(new ReportingUserAgent("ip", Optional.empty()));
+    }
+
+    @Test
+    public void parseStrictShouldThrowOnMissingIp() {
+        expectedException.expect(IllegalArgumentException.class);
+
+        System.out.println(strict.parse("  ; product"));
```
Comment (mbaechler) [inline mdn/src/test/java/org/apache/james/mdn/parsing/ReportingUserAgentParserTest.java]: system.out spotted

File: mdn/src/main/java/org/apache/james/mdn/action/mode/DispositionActionMode.java [hunk 15]
```diff
@@ -34,7 +34,7 @@
 
     public static Optional<DispositionActionMode> fromString(String value) {
         return Stream.of(values())
-            .filter(actionMode -> actionMode.getValue().equalsIgnoreCase(value))
+            .filter(actionMode -> actionMode.getValue().equalsIgnoreCase(value.trim()))
```
Comment (mbaechler) [inline mdn/src/main/java/org/apache/james/mdn/action/mode/DispositionActionMode.java]: this change should have a test in DispositionActionModeTest

File: mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+package org.apache.james.mdn.parsing;
+
+import java.util.List;
+import java.util.Optional;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+
+public class Unwrapper {
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/parsing/Unwrapper.java]: I think that's not unwrapping, that's unfolding (for every field/class/method of this commit, sorry)

File: mdn/src/test/java/org/apache/james/mdn/parsing/UnwrapperTest.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+    @Test
+    public void unwrapShouldUnwrapWrappedLines() {
+        assertThat(Unwrapper.unwrap("aaa\r\n bbb\r\nccc\r\n ddd"))
+            .containsExactly("aaa\nbbb", "ccc\nddd");
+    }
+
+    @Test
+    public void unwrapShouldUnwrapComplexString() {
+        assertThat(Unwrapper.unwrap("aaa\r\n aaa\r\nbbb\r\nccc\r\n ccc\r\n ccc\r\nddd\r\n ddd\r\neee"))
+            .containsExactly("aaa\naaa", "bbb", "ccc\nccc\nccc", "ddd\nddd", "eee");
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/parsing/UnwrapperTest.java]: it's not very clear to me why you replace \r\n by \n

Comment (rouazana) [reply mdn/src/test/java/org/apache/james/mdn/parsing/UnwrapperTest.java]: moreover you don't handle HTAB (3.1.1 of the spec)?

File: mdn/src/test/java/org/apache/james/mdn/parsing/UnparsedFieldTest.java [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+
+    private UnparsedField.Factory strictFactory;
+    private UnparsedField.Factory lenientFactory;
+
+    @Before
+    public void setUp() {
+        strictFactory = new UnparsedField.Factory(true);
+        lenientFactory = new UnparsedField.Factory(false);
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/parsing/UnparsedFieldTest.java]: should be an enum, not a boolean

File: mdn/src/test/java/org/apache/james/mdn/parsing/UnparsedFieldTest.java [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+
+    @Test
+    public void strictParseShouldThrowWhenEmpty() {
+        expectedException.expect(IllegalArgumentException.class);
+
+        strictFactory.parse("");
+    }
+
+    @Test
+    public void strictParseShouldThrowWhenNonName() {
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/parsing/UnparsedFieldTest.java]: s/Non/No/ ?

File: mdn/src/test/java/org/apache/james/mdn/parsing/UnparsedFieldTest.java [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+    public void lenientParseShouldAcceptMultilineValue() {
+        assertThat(lenientFactory.parse("a:b\nc"))
+            .contains(new UnparsedField("a", "b\nc"));
+    }
+
+    @Test
+    public void strictParseShouldAcceptMultilineValue() {
+        assertThat(strictFactory.parse("a:b\nc"))
+            .contains(new UnparsedField("a", "b\nc"));
+    }
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/parsing/UnparsedFieldTest.java]: you don't test (nor handle?) `a: b` case (note the space after the name)

File: mdn/src/main/java/org/apache/james/mdn/fields/Gateway.java [hunk 21]
```diff
@@ -53,7 +58,7 @@ public AddressType getNameType() {
         return nameType;
     }
 
-    public Text getName() {
+    public Text getUaName() {
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/fields/Gateway.java]: Ua?

File: mdn/src/main/java/org/apache/james/mdn/parsing/DispositionParser.java (hunk 13 above)
Comment (rouazana) [reply mdn/src/main/java/org/apache/james/mdn/parsing/DispositionParser.java]: anyway I thought we agreed to make only a Lenient parser in a first time. I don't see the interest of a strict parser for this functionality.

File: mdn/src/main/java/org/apache/james/mdn/parsing/DispositionParser.java [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+        String afterActionMode = value.substring(slashPosition + 1);
+
+        int semiColonPosition = afterActionMode.indexOf(';');
+        Optional<DispositionSendingMode> sendingMode = parseSendingMode(afterActionMode, semiColonPosition);
+        String afterSendingMode = afterActionMode.substring(semiColonPosition + 1);
+
+
+        int secondSlashPosition = afterSendingMode.indexOf('/');
+        Optional<DispositionType> type = parseType(afterSendingMode, secondSlashPosition);
+        String afterType = getModifierString(afterSendingMode, secondSlashPosition);
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/parsing/DispositionParser.java]: maybe you are in a rare case where a regexp make things more obvious :)

File: mdn/src/main/java/org/apache/james/mdn/parsing/GatewayParser.java [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+    public Optional<Field> parse(String value) {
+        Preconditions.checkNotNull(value);
+
+        return Optional.of(factory.parse(value))
+            .filter(this::isUsable)
+            .map(typedValue -> new Gateway(typedValue.getType(), Text.fromRawText(typedValue.getValue())));
+    }
+
+    private boolean isUsable(TypedValue typedValue) {
+        return  !typedValue.getValue().trim().isEmpty();
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/parsing/GatewayParser.java]: 2 spaces

File: mdn/src/main/java/org/apache/james/mdn/parsing/ReportingUserAgentParser.java [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+            if (trimedValue.isEmpty()) {
+                if (strict) {
+                    throw new IllegalArgumentException("Expecting the filed not to be only folding white spaces");
+                }
+                return Optional.empty();
+            }
+            return Optional.of(new ReportingUserAgent(trimedValue.substring(semiColonPosition + 1)));
+        }
+        String name = trimedValue.substring(0, semiColonPosition).trim();
+        String product = trimedValue.substring(semiColonPosition + 1).trim();
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/parsing/ReportingUserAgentParser.java]: instead of these substrings (seen multiple times) you can use String.split(value, 2)

File: mdn/src/main/java/org/apache/james/mdn/sending/mode/DispositionSendingMode.java [hunk 25]
```diff
@@ -34,7 +34,7 @@
 
     public static Optional<DispositionSendingMode> fromString(String value) {
         return Stream.of(values())
-            .filter(sendingMode -> sendingMode.getValue().equalsIgnoreCase(value))
+            .filter(sendingMode -> sendingMode.getValue().equalsIgnoreCase(value.trim()))
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/sending/mode/DispositionSendingMode.java]: missing tests?

File: mdn/src/main/java/org/apache/james/mdn/type/DispositionType.java [hunk 26]
```diff
@@ -36,7 +36,7 @@
 
     public static Optional<DispositionType> fromString(String value) {
         return Stream.of(values())
-            .filter(sendingMode -> sendingMode.getValue().equalsIgnoreCase(value))
+            .filter(sendingMode -> sendingMode.getValue().equalsIgnoreCase(value.trim()))
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/type/DispositionType.java]: also?

File: mdn/src/test/java/org/apache/james/mdn/parsing/DispositionParserTest.java [hunk 27]
```diff
@@ -0,0 +1,192 @@
...
+            .contains(Disposition.builder()
+                .actionMode(DispositionActionMode.Automatic)
+                .sendingMode(DispositionSendingMode.Automatic)
+                .type(DispositionType.Processed)
+                .addModifiers(DispositionModifier.Error, DispositionModifier.Failed)
+                .build());
+    }
+
+    @Test
+    public void parseShouldHandleWellWhenOnlyOneModifier() {
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/parsing/DispositionParserTest.java]: Well What? (and in other test names)

File: mdn/src/test/java/org/apache/james/mdn/parsing/ExtensionFieldParserTest.java [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Test
+    public void parseShouldAcceptEmptyValue() {
+        assertThat(parser.parse(""))
+            .contains(new ExtensionField(NAME, ""));
+    }
+
+    @Test
+    public void parseShouldAcceptFoldingWhiteSpaceValue() {
+        assertThat(parser.parse("  "))
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/parsing/ExtensionFieldParserTest.java]: same than trimming?

File: mdn/src/test/java/org/apache/james/mdn/parsing/OriginalRecipientParserTest.java [hunk 29]
```diff
@@ -0,0 +1,128 @@
...
+
+import org.apache.james.mdn.fields.AddressType;
+import org.apache.james.mdn.fields.OriginalRecipient;
+import org.apache.james.mdn.fields.Text;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+public class OriginalRecipientParserTest {
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/parsing/OriginalRecipientParserTest.java]: you have a different parser for originalRecipient than for finalRecipient?

File: mdn/src/main/java/org/apache/james/mdn/parsing/FieldFinder.java [hunk 30]
```diff
@@ -0,0 +1,137 @@
...
+        Preconditions.checkNotNull(fields);
+
+        this.strict = strict;
+        this.fields = ImmutableList.copyOf(fields);
+    }
+
+    public List<ExtensionField> retrieveExtensions() {
+        return fields.stream()
+            .filter(field -> !RFC_DEFINED_FIELDS.contains(field.getFieldName()))
+            .map(field -> (ExtensionField) field)
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/parsing/FieldFinder.java]: why don't you filter on field's class then?

File: mdn/src/main/java/org/apache/james/mdn/parsing/FieldFinder.java [hunk 31]
```diff
@@ -0,0 +1,137 @@
...
+            .filter(field -> field.getFieldName().equals(fieldName))
+            .collect(Guavate.toImmutableList());
+
+        if (matchingFields.size() > 1) {
+            if (strict) {
+                throw new IllegalStateException(fieldName + " is duplicated.");
+            }
+        }
+
+        return  matchingFields.stream().findFirst();
```
Comment (rouazana) [inline mdn/src/main/java/org/apache/james/mdn/parsing/FieldFinder.java]: 2 spaces

File: mdn/src/test/java/org/apache/james/mdn/MDNReportParserTest.java [hunk 32]
```diff
@@ -0,0 +1,188 @@
...
+                    .actionMode(DispositionActionMode.Automatic)
+                    .sendingMode(DispositionSendingMode.Automatic)
+                    .type(DispositionType.Processed)
+                    .build())
+                .finalRecipientField(new FinalRecipient(Text.fromRawText("final_recipient")))
+                .build());
+    }
+
+    @Test
+    public void parseLenientShouldThrowOnDuplicatedFields() {
```
Comment (rouazana) [inline mdn/src/test/java/org/apache/james/mdn/MDNReportParserTest.java]: s/Throw/TakeFirst/ ?
